### PR TITLE
Spack cache change, bump versions

### PR DESCRIPTION
**Pull request by brenthuisman**
- It's actually a Github cache trick: https://github.com/actions/cache/issues/342#issuecomment-673371329
  - First run: 1h 7m 38s 
  - Second run: same duration. Hmm....
  - Third run: ~10 minutes!
  - 4th: ~7.5 min.
- Bump in configs: macos 10.15 will disappear in a month: https://github.com/actions/virtual-environments/issues/5583
  - Bump clang-max to 14, gcc-max to 11
  - macos-min to 11, macos-max to 12.

### CODE REVIEW

File: .github/workflows/ciwheel.yml [hunk 1]
```diff
@@ -67,8 +67,6 @@ jobs:
     runs-on: ubuntu-latest
 
     steps:
-      - name: Set up Python
-        uses: actions/setup-python@v2
```
Comment (thorstenhater) [inline .github/workflows/ciwheel.yml:71]: Why is this removed? I'd feel more comfortable with explicit versions everywhere.

**Review by thorstenhater — CHANGES_REQUESTED** [review]
A few questions and a more general comment: this is actually mainly bumping versions
and thus mixes two concerns.

File: .github/workflows/lint.yml [hunk 2]
```diff
@@ -12,7 +12,7 @@ jobs:
       fail-fast: false
     steps:
       - name: Set up Python
-        uses: actions/setup-python@v2
+        uses: actions/setup-python@v4
         with:
           python-version: 3.6
```
[the remark this answers is not in the document]
Comment (brenthuisman) [reply .github/workflows/lint.yml:17]: Python min-ver is 3.7, so yes, that can be updated.

Macos wheels are 3.8 and up, because that's the first version that support 'fat wheels'. People who use 3.7 should still be able to build themselves.

File: .github/workflows/spack.yml [hunk 3]
```diff
@@ -23,8 +23,8 @@ jobs:
         uses: actions/cache@v2
         with:
           path: ~/.spack-cache
-          key: cache-${{ github.sha }}
-          restore-keys: cache-
+          key: ccache-${{ github.run_id }}
```
[the remark this answers is not in the document]
Comment (brenthuisman) [reply .github/workflows/spack.yml:26]: no reason.

File: .github/workflows/ciwheel.yml (hunk 1 above)
Comment (brenthuisman) [reply .github/workflows/ciwheel.yml:71]: I removed it everywhere where version should not matter.

File: .github/workflows/spack.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (brenthuisman) [reply .github/workflows/spack.yml:26]: Interestingly it does matter. The commit where I removed it failed, adding back it passed. I can think of no other reason that that the action in fact uses `ccache`, because what else would you use :)

[the remark this answers is not in the document]
Comment (thorstenhater) [reply .github/workflows/spack.yml:26]: That's very obtuse. Please make a big, red *flashing sign* for that.

[the remark this answers is not in the document]
Comment (brenthuisman) [reply .github/workflows/spack.yml]: Dug a little, and could not find any reason to believe it actually matters. Probably the cache was just cold after deleting a `c`. I changed it to sth that assumes no magic, and trust that it'll work. It'll take a run or two extra if you want to be absolutely sure.

**Review by thorstenhater — APPROVED** [review]
👍

File: .github/workflows/spack.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacksongoode) [reply .github/workflows/spack.yml]: There's no need for the double "c" :)
